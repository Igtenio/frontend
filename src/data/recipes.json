[
  {
    "id": 1,
    "name": "Thanks, Obama",
    "user": "JosefBud",
    "userId": 5420,
    "date": "March 20 2019, 16:39:00",
    "notes": "IMPORTANT: CAP Lemon Lime's gravity (g per ml) by default is way above what it should be. If you mix by weight, mix CAP Lemon Lime as 1g = 1ml.\n \n CAP Super Sweet is optional. \n \n VapeWild doesn't add cooling, but 2 drops per 30ml of WS-23 works well if you're into that.\n \n No steep required, shake and vape!",
    "shakeNVape": true,
    "steepTime": 0,
    "maxvg": false,
    "pg": 30,
    "vg": 70,
    "flavors": [
      {
        "vendor": "Flavor West",
        "abbreviation": "FW",
        "name": "Blue Raspberry",
        "id": 1,
        "percent": 3
      },
      {
        "vendor": "The Flavor Apprentice",
        "abbreviation": "TFA",
        "name": "Cherry Extract",
        "id": 2,
        "percent": 1.5
      },
      {
        "vendor": "Capella",
        "abbreviation": "CAP",
        "name": "Lemon Lime",
        "id": 3,
        "percent": 4
      },
      {
        "vendor": "Capella",
        "abbreviation": "CAP",
        "name": "Super Sweet",
        "id": 4,
        "percent": 0.5
      }
    ]
  },
  {
    "id": 2,
    "name": "DaMomma's RY4 Delight",
    "user": "Alisa",
    "userId": 1024,
    "date": "May 26 2016, 08:10:00",
    "notes": "I am really enjoying this one. The Catalan Cream just works really well with RY4 too.\n \n Aged only a few days and it's YUMMY.",
    "shakeNVape": false,
    "steepTime": 4,
    "maxvg": false,
    "pg": 30,
    "vg": 70,
    "flavors": [
      {
        "vendor": "FlavourArt",
        "abbreviation": "FA",
        "name": "Catalan Cream",
        "id": 5,
        "percent": 1
      },
      {
        "vendor": "The Flavor Apprentice",
        "abbreviation": "TFA",
        "name": "RY4 Double",
        "id": 6,
        "percent": 8
      },
      {
        "vendor": "Flavor West",
        "abbreviation": "FW",
        "name": "Sweet Cream",
        "id": 7,
        "percent": 2.5
      },
      {
        "vendor": "Capella",
        "abbreviation": "CAP",
        "name": "Vanilla Custard v2",
        "id": 8,
        "percent": 2
      }
    ]
  },
  {
    "id": 3,
    "name": "Strawberry Fog",
    "user": "spectrometer",
    "userId": 5823,
    "date": "August 14 2015, 23:09:00",
    "notes": "",
    "shakeNVape": true,
    "steepTime": 0,
    "maxvg": true,
    "pg": null,
    "vg": null,
    "flavors": [
      {
        "vendor": "The Flavor Apprentice",
        "abbreviation": "TFA",
        "name": "Bavarian Cream",
        "id": 9,
        "percent": 3
      },
      {
        "vendor": "The Flavor Apprentice",
        "abbreviation": "TFA",
        "name": "Cheesecake (Graham Crust)",
        "id": 10,
        "percent": 5
      },
      {
        "vendor": "The Flavor Apprentice",
        "abbreviation": "TFA",
        "name": "Strawberry Ripe",
        "id": 11,
        "percent": 10
      },
      {
        "vendor": "The Flavor Apprentice",
        "abbreviation": "TFA",
        "name": "Vanilla Custard",
        "id": 12,
        "percent": 2
      }
    ]
  },
  {
    "id": 4,
    "name": "Mother's Unicorn Milk",
    "user": "apwroblewski",
    "userId": 9395,
    "date": "March 20 2015, 07:41:00",
    "notes": "",
    "shakeNVape": true,
    "steepTime": 0,
    "maxvg": false,
    "pg": 50,
    "vg": 50,
    "flavors": [
      {
        "vendor": "The Flavor Apprentice",
        "abbreviation": "TFA",
        "name": "Bavarian Cream",
        "id": 9,
        "percent": 4
      },
      {
        "vendor": "The Flavor Apprentice",
        "abbreviation": "TFA",
        "name": "Cheesecake (Graham Crust)",
        "id": 10,
        "percent": 5
      },
      {
        "vendor": "The Flavor Apprentice",
        "abbreviation": "TFA",
        "name": "Strawberry Ripe",
        "id": 11,
        "percent": 10
      },
      {
        "vendor": "The Flavor Apprentice",
        "abbreviation": "TFA",
        "name": "Sweet Cream",
        "id": 13,
        "percent": 4
      },
      {
        "vendor": "The Flavor Apprentice",
        "abbreviation": "TFA",
        "name": "Vanilla Custard",
        "id": 12,
        "percent": 3
      }
    ]
  },
  {
    "id": 5,
    "name": "Ella's Shortbread Biscuit",
    "user": "VHM",
    "userId": 7582,
    "date": "March 26 2016, 07:38:00",
    "notes": "Comforting crumbly buttery vanilla Scottish shortbread, great as a stand alone or as a base for fruits, creams and all sorts of other flavours.\n \n I enjoy this fresh but if you find INW Biscuit too raw the mix really starts to settle after three days getting richer as it ages.\n \n CAP Sugar Cookie and INW Biscuit - they are great cookie bases that pack a punch. Together they work well at mimicking a proper butter shortbread.\n \n AP brings forward the grain of the mix, adding just the right amount of bite and the CAP V.Custard (v.1) keeps the mix from being too dry on the pallet giving that overall buttery feel.\n \n INW Vanilla Shisha added for its inherent bakery notes and general vanilla decadence.\n \n Feeling lazy? Buy the one shot here: \n UK: https://www.chefsflavours.co.uk/vaping-home-mixers/4160-ellas-shortbread-biscuit.html\n USA: http://www.onestopdiyshop.com/product-p/vhmella30.html\n \n Credit: Betamax (Ella Robinson)",
    "shakeNVape": false,
    "steepTime": 14,
    "maxvg": false,
    "pg": 20,
    "vg": 80,
    "flavors": [
      {
        "vendor": "The Flavor Apprentice",
        "abbreviation": "TFA",
        "name": "Acetyl Pyrazine 5%",
        "id": 14,
        "percent": 0.5
      },
      {
        "vendor": "Inawera",
        "abbreviation": "INW",
        "name": "Biscuit",
        "id": 15,
        "percent": 1.5
      },
      {
        "vendor": "Inawera",
        "abbreviation": "INW",
        "name": "Shisha Vanilla",
        "id": 16,
        "percent": 1
      },
      {
        "vendor": "Capella",
        "abbreviation": "CAP",
        "name": "Sugar Cookie",
        "id": 17,
        "percent": 4
      },
      {
        "vendor": "Capella",
        "abbreviation": "CAP",
        "name": "Vanilla Custard",
        "id": 18,
        "percent": 2
      }
    ]
  }
]
