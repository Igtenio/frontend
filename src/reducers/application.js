import { buildActions } from 'utils';

export const types = buildActions('application', [
  'INIT_APP',
  'LOGIN_USER',
  'LOGIN_USER_SUCCESS',
  'LOGIN_USER_FAILURE',
  'LOGOUT_USER',
  'LOGOUT_USER_SUCCESS',
  'LOGOUT_USER_FAILURE',
  'REGISTER_USER',
  'REGISTER_USER_SUCCESS',
  'REGISTER_USER_FAILURE'
]);

const initApp = () => ({
  type: types.INIT_APP
});

const loginUser = (emailAddress, password) => ({
  type: types.LOGIN_USER,
  emailAddress,
  password
});

const loginUserSuccess = user => ({
  type: types.LOGIN_USER_SUCCESS,
  user
});

const loginUserFailure = error => ({
  type: types.LOGIN_USER_FAILURE,
  error
});

const logoutUser = () => ({
  type: types.LOGOUT_USER
});

const logoutUserSuccess = () => ({
  type: types.LOGOUT_USER_SUCCESS
});

const logoutUserFailure = error => ({
  type: types.LOGOUT_USER_FAILURE,
  error
});

const registerUser = details => ({
  type: types.REGISTER_USER,
  details
});

const registerUserSuccess = () => ({
  type: types.REGISTER_USER_SUCCESS
});

const registerUserFailure = error => ({
  type: types.REGISTER_USER_FAILURE,
  error
});

export const actions = {
  initApp,
  loginUser,
  loginUserSuccess,
  loginUserFailure,
  logoutUser,
  logoutUserSuccess,
  logoutUserFailure,
  registerUser,
  registerUserSuccess,
  registerUserFailure
};

export const initialState = {
  loggingIn: false,
  user: null,
  error: null,
  registration: {
    complete: false,
    error: null
  }
};

export const reducer = (state = initialState, action = {}) => {
  switch (action.type) {
    case types.LOGIN_USER:
      return {
        ...state,
        loggingIn: true
      };
    case types.LOGOUT_USER_SUCCESS:
      return {
        ...state,
        user: null
      };
    case types.LOGIN_USER_SUCCESS:
      return {
        ...state,
        loggingIn: false,
        user: action.user
      };
    case types.LOGIN_USER_FAILURE:
    case types.LOGOUT_USER_FAILURE:
      return {
        ...state,
        loggingIn: false,
        error: action.error
      };
    case types.REGISTER_USER_SUCCESS:
      return {
        ...state,
        registration: {
          complete: true,
          error: null
        }
      };
    case types.REGISTER_USER_FAILURE:
      return {
        ...state,
        registration: {
          complete: true,
          error: action.error
        }
      };
    default:
      return state;
  }
};
