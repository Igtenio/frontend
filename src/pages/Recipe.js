import PropTypes from 'prop-types';
import { Helmet } from 'react-helmet';
import queryString from 'query-string';
import React, { Component } from 'react';

import {
  Container,
  Row,
  Col,
  Table,
  Button,
  ButtonGroup,
  Alert,
  Card
} from 'react-bootstrap';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

import flavorStash from '../data/flavorstash.json';
import recipes from '../data/recipes.json';

export default class Recipe extends Component {
  static propTypes = {
    location: PropTypes.object
  };

  constructor(...args) {
    super(...args);

    this.state = {
      flavors: [],
      favorited: false,
      favoriteIcon: ['far', 'heart'],
      alertClass: 'recipe-alert alert-hidden',
      rating: 0
    };

    this.handleFavoriteClick = this.handleFavoriteClick.bind(this);
    this.handleRatingClick = this.handleRatingClick.bind(this);
    this.renderRatingButtons = this.renderRatingButtons.bind(this);
  }

  componentDidMount() {
    this.findRecipe();
    this.compareWithStash();
  }

  handleFavoriteClick() {
    if (this.state.favorited) {
      this.setState({
        favorited: false,
        favoriteIcon: ['far', 'heart'],
        alertClass: 'recipe-alert alert-fade-in'
      });
    } else {
      this.setState({
        favorited: true,
        favoriteIcon: ['fas', 'heart'],
        alertClass: 'recipe-alert alert-fade-in'
      });
    }

    setTimeout(() => {
      this.setState({ alertClass: 'recipe-alert alert-fade-out alert-hidden' });
    }, 2500);
  }

  handleRatingClick(ratingNumber) {
    if (ratingNumber === this.state.rating) {
      this.setState({ rating: 0 });
    } else {
      this.setState({
        rating: ratingNumber
      });
    }
  }

  findRecipe() {
    const query = this.props.location.search;
    const queryValues = queryString.parse(query);
    const pageId = parseFloat(queryValues.id);

    const recipe = recipes.filter(e => e.id === pageId)[0];

    recipe.flavorTotal = 0;
    recipe.flavors.forEach(flavor => {
      recipe.flavorTotal += flavor.percent;
    });

    if (recipe.notes) {
      const newNotes = recipe.notes.split('\n').map((line, i) => {
        return <p key={i}>{line}</p>;
      });

      recipe.notes = newNotes;
    }

    const newTags = recipe.tags.map((tag, i) => {
      return (
        <a className="link--tags" key={i} href="/">
          {tag}
        </a>
      );
    });

    recipe.tags = newTags;

    this.setState(recipe);
  }

  compareWithStash() {
    // eslint-disable-next-line
    console.log(this.state);
    const query = this.props.location.search;
    const queryValues = queryString.parse(query);
    const pageId = parseFloat(queryValues.id);

    const recipe = recipes.filter(e => e.id === pageId)[0];

    let index = 0;

    recipe.flavors.forEach(recipeFlavor => {
      let stashIndex = 0;

      flavorStash.forEach(stashFlavor => {
        if (recipeFlavor.id === stashFlavor.id) {
          recipe.flavors[index].inStash = true;
          return;
        }

        stashIndex++;
        if (stashIndex === flavorStash.length) {
          recipe.flavors[index].inStash = false;
          return;
        }
      });

      index++;
    });

    this.setState(recipe);
  }

  renderRatingButtons(ratingNum) {
    const buttons = [];

    let ratingKey = 1;

    for (let i = 0; i < ratingNum; i++) {
      buttons.push(
        <Button
          className="rating-button button--recipe"
          key={i + 1}
          onClick={() => this.handleRatingClick(i + 1)}
        >
          <span>
            <FontAwesomeIcon icon={['fas', 'star']} />
          </span>
        </Button>
      );
      ratingKey++;
    }

    for (let a = ratingNum; a < 5; a++) {
      buttons.push(
        <Button
          className="rating-button button--recipe"
          key={a + 1}
          onClick={() => this.handleRatingClick(a + 1)}
        >
          <span>
            <FontAwesomeIcon icon={['far', 'star']} />
          </span>
        </Button>
      );
      ratingKey++;
    }

    if (ratingKey === 6) {
      return <ButtonGroup>{buttons}</ButtonGroup>;
    }
  }

  render() {
    return (
      <Container className="text-center container--recipe">
        <Helmet title="Recipe" />
        <Row className="justify-content-center">
          <Alert variant="success" className={this.state.alertClass}>
            {this.state.favorited ? (
              <span>Recipe added to your favorites</span>
            ) : (
              <span>Recipe removed from your favorites</span>
            )}
          </Alert>
          <Col xs="auto">
            <Button className="button-animation button--recipe">
              <span>Make this recipe</span>
            </Button>
          </Col>
          <Col xs="auto">
            <Button
              onClick={this.handleFavoriteClick}
              className="button--favorite button--recipe"
            >
              <span>
                <FontAwesomeIcon icon={this.state.favoriteIcon} />
              </span>
            </Button>
          </Col>
          <Col xs="auto">{this.renderRatingButtons(this.state.rating)}</Col>
        </Row>
        <hr />
        <Row>
          <Col lg={{ span: 8, offset: 2 }} xs={{ span: 12 }}>
            <Card className="text-center">
              <Card.Header>{this.state.name}</Card.Header>
              <Card.Body>
                <Row>
                  <Col md={{ span: 5 }} xs={{ span: 12 }}>
                    <img
                      src="/media/card-test-1.jpg"
                      alt="card test"
                      className="w-75"
                    />
                  </Col>
                  <Col md={{ span: 7 }} xs={{ span: 12 }}>
                    <Card.Title>Version {this.state.version}</Card.Title>
                    <Card.Title>Notes</Card.Title>
                    <Card.Text>
                      {this.state.notes ? (
                        <div>{this.state.notes}</div>
                      ) : (
                        <div>
                          This user did not enter any notes for this recipe.
                        </div>
                      )}
                    </Card.Text>
                  </Col>
                </Row>
              </Card.Body>
              <Card.Footer className="text-muted">
                <span>
                  Created by{' '}
                  <a href={'/user?id=' + this.state.userId}>
                    {this.state.user}
                  </a>{' '}
                  on {this.state.date}
                </span>
              </Card.Footer>
            </Card>
          </Col>
        </Row>
        <Row>
          <Col lg={{ span: 8, offset: 2 }} xs={{ span: 12 }}>
            <Table striped bordered>
              <caption className="text-center">
                {this.state.maxvg ? (
                  <span>Max VG |&nbsp;</span>
                ) : (
                  <span>
                    {this.state.pg}% PG / {this.state.vg}% VG |&nbsp;
                  </span>
                )}
                {this.state.shakeNVape ? (
                  <span>Shake & Vape |&nbsp;</span>
                ) : (
                  <span>Steep for {this.state.steepTime} days |&nbsp;</span>
                )}
                <span>Flavor total: {this.state.flavorTotal}%</span>
              </caption>
              <thead>
                <tr>
                  <th>Flavor</th>
                  <th>Percent</th>
                  <th>Stash</th>
                </tr>
              </thead>
              <tbody>
                {this.state.flavors.map((flavor, index) => (
                  <tr key={index}>
                    <td>
                      <a href={'/flavor?id=' + flavor.id}>
                        {flavor.abbreviation} {flavor.name}
                      </a>
                    </td>
                    <td>{flavor.percent.toFixed(1)}%</td>
                    <td>
                      {flavor.inStash ? (
                        <FontAwesomeIcon icon={['far', 'check-square']} />
                      ) : (
                        <FontAwesomeIcon icon={['far', 'square']} />
                      )}
                    </td>
                  </tr>
                ))}
              </tbody>
            </Table>
          </Col>
        </Row>
        <Row>
          <Col lg={{ span: 6, offset: 3 }} xs={{ span: 12 }}>
            <h3>Tags:</h3>
            <div className="d-flex justify-content-center flex-wrap">
              {this.state.tags}
            </div>
          </Col>
        </Row>
      </Container>
    );
  }
}
