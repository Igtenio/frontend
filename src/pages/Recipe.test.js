import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';

import { initialState } from 'reducers/application';
import { Recipe } from './Recipe';
import recipes from '../data/recipes.json';

const mockStore = configureStore();
const store = mockStore({ application: initialState });

describe('<Recipe />', () => {
  const actions = {
    registerUser: jest.fn()
  };
  const props = search => ({
    location: {
      search
    }
  });

  it('renders correctly', () => {
    const component = renderer.create(
      <Provider store={store}>
        <Recipe {...props('?id=1')} />
      </Provider>
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('can findRecipe', () => {
    const component = renderer.create(
      <Recipe actions={actions} {...props('?id=1')} />
    );
    const { instance } = component.root.findByType(Recipe);

    expect(instance).toBeDefined();
    instance.findRecipe();
    expect(instance.state).toEqual({
      ...recipes[0],
      flavorTotal: 9,
      favoriteIcon: ['far', 'heart'],
      favorited: false,
      rating: 0
    });
  });
});
