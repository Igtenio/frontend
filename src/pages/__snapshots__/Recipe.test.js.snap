// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Recipe /> renders correctly 1`] = `
<div
  className="text-center container"
>
  <div
    className="row"
  >
    <div
      className="col-md-2 offset-md-3"
    >
      <img
        alt="card test"
        className="w-100 img-thumbnail"
        src="/media/card-test-1.jpg"
      />
    </div>
    <div
      className="col-md-4"
    >
      <h1
        className="recipeTitle"
      >
        Thanks, Obama
      </h1>
      <h2>
        Created by
         
        <a
          href="/user?id=5420"
        >
          JosefBud
        </a>
      </h2>
      <p>
        on 
        March 20 2019, 16:39:00
      </p>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col-md-6 offset-md-3"
    >
      <table
        className="table table-striped table-bordered table-hover"
      >
        <thead>
          <tr>
            <th>
              Flavor
            </th>
            <th>
              Percent
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              <a
                href="/flavor?id=1"
              >
                FW
                 
                Blue Raspberry
              </a>
            </td>
            <td>
              3.0
              %
            </td>
          </tr>
          <tr>
            <td>
              <a
                href="/flavor?id=2"
              >
                TFA
                 
                Cherry Extract
              </a>
            </td>
            <td>
              1.5
              %
            </td>
          </tr>
          <tr>
            <td>
              <a
                href="/flavor?id=3"
              >
                CAP
                 
                Lemon Lime
              </a>
            </td>
            <td>
              4.0
              %
            </td>
          </tr>
          <tr>
            <td>
              <a
                href="/flavor?id=4"
              >
                CAP
                 
                Super Sweet
              </a>
            </td>
            <td>
              0.5
              %
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col"
    >
      <h3>
        30
        % PG / 
        70
        % VG
      </h3>
      <h3>
        Shake & Vape
      </h3>
      <h3>
        Flavor total: 
        9
        %
      </h3>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col"
    >
      <h2>
        Notes
      </h2>
      <p>
        IMPORTANT: CAP Lemon Lime's gravity (g per ml) by default is way above what it should be. If you mix by weight, mix CAP Lemon Lime as 1g = 1ml.
 
 CAP Super Sweet is optional. 
 
 VapeWild doesn't add cooling, but 2 drops per 30ml of WS-23 works well if you're into that.
 
 No steep required, shake and vape!
      </p>
    </div>
  </div>
</div>
`;
